File: crates/asterism-ui/src/lib/material-layer.ts
// material-layer — the reading of a material's chapter bands, separated
// from the component that renders them.
//
// A band (`MaterialLayer`) is one reading of how a material is divided,
// and it says who produced that reading: the file's own declaration, a
// person's, or a job's. The chapters hang off the band rather than off
// the asset, which is what lets re-reading a file replace its own list
// without touching the one a person wrote.
//
// Four facts of that wire shape decide almost everything below, and each
// is easy to get wrong in a way no type catches:
//
//   * **A band has no display name.** It is described by what it *is* —
//     `(origin, role)` — and a surface derives a caption from the pair
//     (`material_layer.rs`, "No display name"). A caption stored beside
//     the pair would be the half that drifts.
//   * **A chapter's `id` is not stable across a re-scan.** Re-reading a
//     file replaces an imported band's rows wholesale, so the row a
//     reader was looking at comes back with a new id. `(layer_id, ord)`
//     is what names the same section before and after, and it is what
//     the `{#each}` keys on.
//   * **`end_ms: null` means the section states no end**, not "runs to
//     the end of the media" (`chapter_mark.rs` / `TimelineSpan`). It is
//     rendered as an absent end, never as a computed one.
//   * **Only a `user` band may be written into.** The service refuses a
//     hand edit of an imported or machine band
//     (`material_layer_service::require_user_owned`), so the surface
//     asks the same question before offering the affordance — an absent
//     button is a better answer than a rejected write.
//
// These live here rather than in `MaterialChapters.svelte` for the
// reason `material-mark.ts` gives: vitest runs on Node and cannot mount
// a component (`vite.config.ts` `test.environment: "node"`), and the
// command builders are one half of a two-sided contract whose other half
// is Rust — a test that pins the shape this side sends is the only thing
// that fails when the two drift.
import { fmtDurationMs } from "./formatters";
import type {
  ChapterMarkDto,
  CreateMaterialLayerCommand,
  EditChapterMarkCommand,
  MaterialLayerViewDto,
  PostChapterMarkCommand,
} from "../bindings";

/// The role slug of a band that holds chapters. The other role
/// (`annotation`) holds notes and is read through `list_material_marks`
/// by `MaterialMarks.svelte`; this surface never shows one.
export const STRUCTURE_ROLE = "structure";

/// The origin slug of a band a person owns — the only one a hand edit
/// may touch.
export const USER_ORIGIN = "user";

/// The bands of an asset that hold chapters, in the order the backend
/// handed them over.
///
/// `list_material_layers` returns every band over the material,
/// annotation ones included (their `chapters` is always empty). Filtering
/// by role here rather than asking the backend for a subset keeps the one
/// call that the panel needs: the bands to choose between and the
/// contents of the chosen one arrive at the same moment.
///
/// The order is the backend's (`ord`, then id). Re-sorting would put a
/// second opinion about display order in the client.
export function structureBands(
  views: readonly MaterialLayerViewDto[],
): MaterialLayerViewDto[] {
  return views.filter((v) => v.layer.role === STRUCTURE_ROLE);
}

/// The caption for a band, derived from the `(origin, role)` pair.
///
/// Total over the wire, not over the enum: `origin` and `role` arrive as
/// strings, and a build that reads a row written by a newer one gets a
/// slug it has no variant for. That case falls through to the pair
/// itself, which is at least true, rather than to a caption that claims
/// the band is something it is not.
export function bandLabel(origin: string, role: string): string {
  if (role === STRUCTURE_ROLE) {
    switch (origin) {
      case "imported":
        return "From the file";
      case USER_ORIGIN:
        return "Yours";
      case "machine":
        return "Detected";
    }
  } else if (role === "annotation") {
    switch (origin) {
      case "imported":
        return "Notes from the file";
      case USER_ORIGIN:
        return "Your notes";
      case "machine":
        return "Detected notes";
    }
  }
  return `${origin} ${role}`;
}

/// Whether a person may write into a band of this origin.
///
/// Mirrors `require_user_owned`: imported bands are replaced by reading
/// the file again and machine bands by re-running the job, so neither
/// takes a hand edit.
export function bandEditable(origin: string): boolean {
  return origin === USER_ORIGIN;
}

/// Which band the panel should show: the one already open if it is still
/// there, otherwise the default one, otherwise the first, otherwise
/// nothing.
///
/// Keeping the current choice across a reload is what makes "add a
/// chapter" not throw the reader back to the file's band. The default is
/// only consulted when the current id is gone — which happens when the
/// band was deleted, or when this is the first load and there is no
/// current id at all.
export function pickBandId(
  bands: readonly MaterialLayerViewDto[],
  current: string | null,
): string | null {
  if (current !== null && bands.some((b) => b.layer.id === current)) return current;
  const preferred = bands.find((b) => b.layer.is_default) ?? bands[0];
  return preferred ? preferred.layer.id : null;
}

/// The keys an `{#each}` over a band's chapters should use, positionally.
///
/// **Not the chapter's `id`.** Re-reading a material replaces an imported
/// band's rows, so the same section arrives with a different id and a
/// keyed block would tear down and rebuild every row — losing focus and
/// scroll for a list whose content did not change. `(layer_id, ord)` is
/// the pair that names the same section across that replacement.
///
/// Returns the whole list rather than one key at a time because that
/// pair is **not unique**, and a Svelte key must be. `ord` is the band's
/// stated reading order, but nothing enforces distinctness: the schema's
/// `idx_chapter_mark_layer_ord` is a plain index, the column carries
/// `DEFAULT 0`, and a container is free to declare its sections without
/// one — so a whole band can arrive as `ord = 0`. A repeated key is not
/// a doubled row: Svelte answers `each_key_duplicate` by throwing while
/// reconciling, which takes down the pane the list sits in. That is the
/// failure `DetailPane.test.ts` was written for after a row carried the
/// same label twice on 2026-07-20, and this is the same shape one layer
/// down.
///
/// Repeats are therefore disambiguated by their position *among the
/// repeats* — the first `layer:0` stays `layer:0`, the second becomes
/// `layer:0#1`. That keeps the property the pair was chosen for: the
/// suffix is a function of the order the band states, not of the ids,
/// so a re-scan that renumbers every `id` still produces the same key
/// list for the same sections.
export function chapterRowKeys(chapters: readonly ChapterMarkDto[]): string[] {
  const seen = new Map<string, number>();
  return chapters.map((c) => {
    const base = `${c.layer_id}:${c.ord}`;
    const nth = seen.get(base) ?? 0;
    seen.set(base, nth + 1);
    return nth === 0 ? base : `${base}#${nth}`;
  });
}

/// The `ord` a chapter appended to this band should carry.
///
/// One past the highest in the band, rather than the row count: a band
/// whose middle row was deleted has a gap, and reusing an `ord` that is
/// still in the band would make two rows share a key.
export function nextChapterOrd(chapters: readonly ChapterMarkDto[]): number {
  let max = -1;
  for (const c of chapters) {
    if (Number.isFinite(c.ord) && c.ord > max) max = c.ord;
  }
  return max + 1;
}

/// Builds the command that opens a band of one's own over an asset's
/// primary material.
///
/// `material_ord: null` names the primary original — the axis
/// `asset.duration_ms` measures and `HTMLMediaElement.currentTime`
/// reports (`material_layer.rs` `PRIMARY_MATERIAL_ORD`). The pane plays
/// that one, so it is the one it marks.
///
/// `ord` is one past the bands already there, for the same reason
/// `nextChapterOrd` is: it decides display order, and a repeat would
/// leave two bands with no stated order between them.
export function buildCreateBandCommand(
  assetId: string,
  bands: readonly MaterialLayerViewDto[],
): CreateMaterialLayerCommand {
  let max = -1;
  for (const b of bands) {
    if (Number.isFinite(b.layer.ord) && b.layer.ord > max) max = b.layer.ord;
  }
  return {
    asset_id: assetId,
    material_ord: null,
    role: STRUCTURE_ROLE,
    ord: max + 1,
  };
}

/// Builds the command that adds a section at the playhead, or `null`
/// when there is nothing to add.
///
/// `end_ms: null` is a section that states no end, which is what the
/// panel can honestly say: the person marked where a section *starts*,
/// and nothing was said about where it stops. An end computed from the
/// next chapter would be the panel inventing a claim the person did not
/// make.
///
/// An empty label is legal here, unlike a mark's body — a container is
/// free to declare an untitled section and a person is free to write
/// one (`chapter_mark.rs` `validate`). What is refused is a position
/// that is not one.
export function buildPostChapterCommand(
  layerId: string,
  startMs: number,
  label: string,
  chapters: readonly ChapterMarkDto[],
): PostChapterMarkCommand | null {
  if (!Number.isFinite(startMs) || startMs < 0) return null;
  return {
    layer_id: layerId,
    start_ms: Math.round(startMs),
    end_ms: null,
    label: label.trim(),
    ord: nextChapterOrd(chapters),
  };
}

/// Builds the command that retitles a section, or `null` when the title
/// is the one it already has.
///
/// `start_ms: null` is the load-bearing field: it means "leave the
/// section where it is". Sending the stored start back instead would
/// travel through the same arm that a move does and re-validate a span
/// nobody asked to change — and, because `end_ms` is only read when
/// `start_ms` is present, would quietly restate the end as well
/// (`material_layer_service::edit_chapter_mark`).
export function buildRenameCommand(
  chapter: ChapterMarkDto,
  label: string,
): EditChapterMarkCommand | null {
  const trimmed = label.trim();
  if (trimmed === chapter.label) return null;
  return {
    layer_id: chapter.layer_id,
    chapter_id: chapter.id,
    label: trimmed,
    start_ms: null,
    end_ms: null,
    ord: null,
  };
}

/// Builds the command that moves a section to a new start, or `null`
/// when it is already there.
///
/// Carries the label and the end it already had. The label because the
/// field is not optional — an edit always states a title, and the one it
/// states here is the unchanged one. The end because `end_ms` is read
/// only alongside `start_ms`, so omitting it while moving would silently
/// convert a section with a stated end into one without.
///
/// Keeping the end can make the request invalid: dragging a start past
/// its own end is an inverted interval, which `TimelineSpan::new`
/// refuses. That refusal is surfaced rather than avoided — dropping the
/// end here to make the write succeed would discard a stated fact to
/// spare the caller an error message.
export function buildMoveCommand(
  chapter: ChapterMarkDto,
  startMs: number,
): EditChapterMarkCommand | null {
  if (!Number.isFinite(startMs) || startMs < 0) return null;
  const rounded = Math.round(startMs);
  if (rounded === chapter.start_ms) return null;
  return {
    layer_id: chapter.layer_id,
    chapter_id: chapter.id,
    label: chapter.label,
    start_ms: rounded,
    end_ms: chapter.end_ms,
    ord: null,
  };
}

/// How a section's extent reads: its start alone when it states no end,
/// `start – end` when it states one.
///
/// The absent end is printed as absent. `end_ms: null` says the section
/// declares no end, and the two readings a computed end would invite —
/// "runs to the next chapter", "runs to the end of the media" — are both
/// claims the data does not make.
export function chapterRangeLabel(startMs: number, endMs: number | null): string {
  const start = fmtDurationMs(startMs);
  return endMs === null ? start : `${start} – ${fmtDurationMs(endMs)}`;
}

/// The sentence to show in place of a chapter list, or `null` when there
/// are chapters to show.
///
/// The two empty cases are different facts and are never given the same
/// words:
///
///   * **No structure band at all** — nobody has looked. The material
///     has not been read for chapters (or was imported before it was
///     read for them), and what the file declares is still unknown.
///   * **A band holding no chapters** — somebody looked and the answer
///     was "none". For an imported band that is the file's own statement
///     that it declares no divisions; for a band of one's own it is
///     simply an empty list waiting to be written.
///
/// Collapsing the two into "No chapters" would report an unasked
/// question as an answered one.
export function chapterListNote(
  bands: readonly MaterialLayerViewDto[],
  active: MaterialLayerViewDto | null,
): string | null {
  if (bands.length === 0) return "This material has not been read for chapters.";
  if (active === null) return "This material has not been read for chapters.";
  if (active.chapters.length > 0) return null;
  return active.layer.origin === USER_ORIGIN
    ? "No sections in this band yet."
    : "This file declares no chapters.";
}

File: crates/asterism-core/src/domain/material_layer.rs
//! `MaterialLayer` — one band of marks over an Asset's material, and
//! the thing that says **where a mark came from**.
//!
//! A material can be marked more than once by more than one hand. The
//! container itself declares chapters; a person disagrees with them and
//! writes their own; a job derives a third set from the audio. Before
//! this type those three were the same rows, distinguishable only by
//! whoever happened to have written them last — so "re-read the file"
//! either destroyed a person's work or duplicated the file's, and there
//! was no answer to "which of these did I write?".
//!
//! The layer carries that answer as data. [`LayerOrigin`] says who
//! produced the band, [`LayerRole`] says what kind of thing it holds,
//! and marks belong to a layer rather than directly to the asset:
//! [`ChapterMark`](crate::domain::chapter_mark) hangs off a `Structure`
//! layer, [`MaterialMark`](crate::domain::material_mark) off an
//! `Annotation` one.
//!
//! Design notes:
//!
//! - **The layer is addressed through the asset, not the material.**
//!   `(asset_id, material_ord)` names the original the band is over,
//!   the same pair
//!   [`Material`](crate::domain::material) is identified by, and the
//!   same one `MaterialMark` resolves to `ord == 0` by convention.
//!   Materials are aggregate-internal, so there is no material id to
//!   reference and this is the shape the aggregate offers.
//! - **`Imported` is immutable, and that rule is not here.** Whether a
//!   caller may write into a band depends on which caller it is — a
//!   person editing, or the re-probe job replacing the file's own
//!   declaration wholesale — and the entity cannot see which. The
//!   application layer holds it
//!   ([`material_layer_service`](crate::application::material_layer_service)),
//!   in the one place both routes pass through.
//! - **"The default band" is a cross-row fact**, so it is not enforced
//!   here. At most one layer per `(asset, material_ord, role)` carries
//!   `is_default`, and that is a partial unique index in the schema (see
//!   the V78 doc comment in `migrations.rs`): a rule about *other rows*
//!   cannot be checked by a value holding one of them, and a check that
//!   read them would be a race between its read and its write.
//!   [`Self::validate`] carries the half that is self-contained.
//! - **No display name.** A band is described by what it *is* —
//!   `(origin, role)` — and a surface renders that pair. Storing a
//!   caption as well would make "the imported layer" and whatever the
//!   caption says two answers to one question, and the caption would be
//!   the one that drifts.

use crate::domain::value::{AssetId, MaterialLayerId};
use crate::error::DomainError;

/// The `material_ord` of an asset's primary original — the axis
/// `asset.duration_ms` measures and `HTMLMediaElement.currentTime`
/// reports.
///
/// Named rather than written as a bare `0` at each call site: the
/// surfaces that mark a timeline
/// ([`MaterialMark`](crate::domain::material_mark) and everything above
/// it) address the primary original by convention, and a `0` in an
/// argument list next to another ordinal says nothing about which
/// convention it is.
pub const PRIMARY_MATERIAL_ORD: u32 = 0;

/// Who produced a layer's contents.
///
/// The axis that decides whether a person may edit the band. It is a
/// closed enum rather than an open slug (unlike
/// [`Modality`](crate::domain::value::Modality)) because each value
/// carries a *rule* — an imported band is replaced by re-reading the
/// file, a user band is edited by hand — and a value nobody has written
/// a rule for is a band nothing knows how to write into.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LayerOrigin {
    /// Read out of the material itself: the chapter list the container
    /// declares, an embedded track. Replaced wholesale when the
    /// material is read again; never edited in place.
    Imported,
    /// Written by the person using Asterism. The only origin a hand
    /// edit may touch.
    User,
    /// Derived by a job from the material's content (a scene cut, a
    /// silence split). Machine-owned like `Imported`, but produced here
    /// rather than found in the file.
    Machine,
}

impl LayerOrigin {
    /// Every variant, in the order the `CHECK (origin IN (...))` in the
    /// schema lists them.
    ///
    /// Exists so the round trip below can be proved over the whole enum
    /// rather than over the three values someone remembered to write in
    /// a test.
    pub const ALL: &'static [Self] = &[Self::Imported, Self::User, Self::Machine];

    /// Slug stored in the `origin` column and used on the wire.
    ///
    /// One spelling, so the adapter, the contract layer and the DDL's
    /// `CHECK` cannot disagree about what a band's origin is called.
    pub fn slug(&self) -> &'static str {
        match self {
            Self::Imported => "imported",
            Self::User => "user",
            Self::Machine => "machine",
        }
    }

    /// Reads a slug back. `None` for anything this build has no variant
    /// for.
    ///
    /// Deliberately not `Result`: the two callers want different errors
    /// out of the same failure. A row holding an unknown slug is an
    /// infrastructure fact (`DomainError::Infra`); a command carrying
    /// one is a caller error (`DomainError::Validation`). Returning
    /// `Option` lets each say so in its own words instead of one of them
    /// restating the other's.
    pub fn from_slug(slug: &str) -> Option<Self> {
        Self::ALL.iter().copied().find(|o| o.slug() == slug)
    }
}

/// What kind of marks a layer holds.
///
/// Not a property of the marks themselves: the same timeline position
/// means a different thing in a chapter list ("this section starts
/// here") than in a note ("look at this"). The role is what the two
/// aggregates hang off, and what keeps a `set_default` on one from
/// moving the other.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LayerRole {
    /// A reading of how the material is *divided*: chapters. Holds
    /// [`ChapterMark`](crate::domain::chapter_mark).
    Structure,
    /// Notes fastened to positions in the material. Holds
    /// [`MaterialMark`](crate::domain::material_mark).
    Annotation,
}

impl LayerRole {
    /// Every variant, in the order the `CHECK (role IN (...))` in the
    /// schema lists them.
    pub const ALL: &'static [Self] = &[Self::Structure, Self::Annotation];

    /// Slug stored in the `role` column and used on the wire.
    pub fn slug(&self) -> &'static str {
        match self {
            Self::Structure => "structure",
            Self::Annotation => "annotation",
        }
    }

    /// Reads a slug back; `None` for anything this build has no variant
    /// for. Same two-caller reasoning as [`LayerOrigin::from_slug`].
    pub fn from_slug(slug: &str) -> Option<Self> {
        Self::ALL.iter().copied().find(|r| r.slug() == slug)
    }
}

/// One band of marks over an Asset's material.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MaterialLayer {
    /// Surrogate id (UUID v7; layers are read in `ord` order, so the
    /// embedded timestamp only serves as a tie-break).
    pub id: MaterialLayerId,
    /// Asset whose material the band is over.
    pub asset_id: AssetId,
    /// Which of that asset's originals (`0` = the primary one, the axis
    /// `asset.duration_ms` measures).
    pub material_ord: u32,
    /// Who produced the contents.
    pub origin: LayerOrigin,
    /// What kind of marks it holds.
    pub role: LayerRole,
    /// Whether this is the band a surface shows, and the one a new mark
    /// lands in, when the caller names no other. At most one per
    /// `(asset_id, material_ord, role)` — enforced by the schema, not
    /// here (see the module doc).
    pub is_default: bool,
    /// Display order within `(asset_id, material_ord, role)`.
    pub ord: u32,
}

impl MaterialLayer {
    /// Opens a band over `asset_id`'s material.
    pub fn new(
        asset_id: AssetId,
        material_ord: u32,
        origin: LayerOrigin,
        role: LayerRole,
        is_default: bool,
        ord: u32,
    ) -> Result<Self, DomainError> {
        let layer = Self {
            id: MaterialLayerId::new(),
            asset_id,
            material_ord,
            origin,
            role,
            is_default,
            ord,
        };
        layer.validate()?;
        Ok(layer)
    }

    /// Rebuilds a layer read back from storage.
    ///
    /// Adapters route every row through here rather than assembling the
    /// struct field by field, for the reason
    /// [`MaterialMark::rehydrate`](crate::domain::material_mark::MaterialMark::rehydrate)
    /// gives: this is the read door, and it is the only thing between a
    /// row that arrived some other way — a hand-written `INSERT`, a
    /// migration, a build that had a variant this one does not — and a
    /// caller holding it as a valid layer.
    ///
    /// Failures are `Validation`; an adapter restates them as `Infra`,
    /// since a row like this being present is an infrastructure fact
    /// rather than a caller error.
    pub fn rehydrate(
        id: MaterialLayerId,
        asset_id: AssetId,
        material_ord: u32,
        origin: LayerOrigin,
        role: LayerRole,
        is_default: bool,
        ord: u32,
    ) -> Result<Self, DomainError> {
        let layer = Self {
            id,
            asset_id,
            material_ord,
            origin,
            role,
            is_default,
            ord,
        };
        layer.validate()?;
        Ok(layer)
    }

    /// Checks the one rule this value can hold on its own: **the
    /// default annotation band is the user's**.
    ///
    /// Posting a note names no layer — a person clicking a timeline is
    /// not choosing a band — so the service resolves the default
    /// annotation layer and writes into it, creating one if the asset
    /// has none
    /// ([`material_layer_service`](crate::application::material_layer_service)).
    /// If that band could be `Imported` or `Machine`, every note would
    /// land in a band the immutability rule forbids writing to: either
    /// the guard refuses the person's own note, or the write goes
    /// through and the next re-probe deletes it along with the file's
    /// content. Refusing the row is what keeps that state unreachable
    /// in the first place, and the schema mirrors it
    /// (`CHECK (role <> 'annotation' OR is_default = 0 OR origin =
    /// 'user')`, V78).
    ///
    /// Structure bands are deliberately not covered: the chapter list a
    /// reader wants by default is usually the one the file declares, so
    /// `(Imported, Structure, is_default)` is the *expected* row there.
    ///
    /// Uniqueness of the default is a rule about other rows and lives
    /// in the schema (module doc). Every door calls this — `new`,
    /// `rehydrate`, and an adapter's `save` — because the fields are
    /// public, so a record update reaches them without passing a
    /// constructor.
    pub fn validate(&self) -> Result<(), DomainError> {
        if self.role == LayerRole::Annotation && self.is_default && self.origin != LayerOrigin::User
        {
            return Err(DomainError::Validation(format!(
                "the default annotation layer is where a new mark lands, so it must be \
                 written by the user; origin {:?} cannot carry it",
                self.origin
            )));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Slugs round-trip, over the whole enum rather than over a
    /// hand-written list.
    ///
    /// `ALL` is what makes that true: a variant added without a slug
    /// fails to compile in `slug`, and one added without a place in
    /// `ALL` is caught by the count assertion — which is the failure a
    /// `for` loop over a literal list would miss.
    #[test]
    fn origin_and_role_slugs_round_trip() {
        assert_eq!(LayerOrigin::ALL.len(), 3, "a new origin belongs in ALL");
        for origin in LayerOrigin::ALL {
            assert_eq!(LayerOrigin::from_slug(origin.slug()), Some(*origin));
        }
        assert_eq!(LayerRole::ALL.len(), 2, "a new role belongs in ALL");
        for role in LayerRole::ALL {
            assert_eq!(LayerRole::from_slug(role.slug()), Some(*role));
        }

        // The literal spellings, because the schema's CHECK names them
        // and nothing else in Rust does.
        assert_eq!(LayerOrigin::Imported.slug(), "imported");
        assert_eq!(LayerOrigin::User.slug(), "user");
        assert_eq!(LayerOrigin::Machine.slug(), "machine");
        assert_eq!(LayerRole::Structure.slug(), "structure");
        assert_eq!(LayerRole::Annotation.slug(), "annotation");

        assert_eq!(LayerOrigin::from_slug("Imported"), None, "case matters");
        assert_eq!(LayerOrigin::from_slug("derived"), None);
        assert_eq!(LayerRole::from_slug(""), None);
    }

    /// The default annotation band has to be the user's — and the three
    /// neighbouring rows that stay legal are asserted beside it.
    ///
    /// Without them the rule would pass with `validate` replaced by
    /// `return Err(...)`: an imported *structure* default is the
    /// ordinary case (the chapter list the file declares), and a
    /// non-default imported annotation band is exactly what a re-probe
    /// writes.
    #[test]
    fn only_the_user_owns_the_default_annotation_band() {
        let asset = AssetId::new();
        let open =
            |origin, role, is_default| MaterialLayer::new(asset, 0, origin, role, is_default, 0);

        assert!(
            open(LayerOrigin::User, LayerRole::Annotation, true).is_ok(),
            "the band a new note lands in"
        );
        assert!(
            open(LayerOrigin::Imported, LayerRole::Structure, true).is_ok(),
            "the file's own chapter list is the expected default there"
        );
        assert!(
            open(LayerOrigin::Imported, LayerRole::Annotation, false).is_ok(),
            "an imported annotation band is fine as long as it is not the default"
        );

        for origin in [LayerOrigin::Imported, LayerOrigin::Machine] {
            let err = open(origin, LayerRole::Annotation, true)
                .expect_err("a note would land in a band nobody may write to");
            assert!(
                matches!(err, DomainError::Validation(_)),
                "expected Validation, got {err:?}"
            );
        }
    }

    /// `rehydrate` applies the same rule the constructor does.
    ///
    /// The fields are public, so the value a row is read into can be
    /// reached by a record update; the read door is what keeps a row
    /// written around the schema from arriving as a valid layer.
    #[test]
    fn rehydrate_rejects_rows_the_constructor_would_refuse() {
        let id = MaterialLayerId::new();
        let asset = AssetId::new();
        assert!(
            MaterialLayer::rehydrate(
                id,
                asset,
                0,
                LayerOrigin::Machine,
                LayerRole::Annotation,
                true,
                0
            )
            .is_err(),
            "a machine-owned default annotation band stays refused on the way back in"
        );
        let ok = MaterialLayer::rehydrate(
            id,
            asset,
            2,
            LayerOrigin::Machine,
            LayerRole::Annotation,
            false,
            7,
        )
        .expect("the same band, not the default, reads back");
        assert_eq!(ok.material_ord, 2);
        assert_eq!(ok.ord, 7);
        assert!(!ok.is_default);
    }
}
